src/herder/HerderImpl.cpp

```diff
@@ -2618,4 +2624,65 @@ HerderImpl::getTx(Hash const& hash) const
     return classic;
 }
 
+std::shared_ptr<MutableTransactionResultBase const>
+HerderImpl::checkValidCached(LedgerSnapshot const& ls,
+                             TransactionFrameBaseConstPtr const& tx,
+                             uint64_t lowerBoundCloseTimeOffset,
+                             uint64_t upperBoundCloseTimeOffset,
+                             DiagnosticEventManager& diagnosticEvents)
+{
+    ZoneScoped;
+    releaseAssert(threadIsMain());
+
+    bool timeDependent = static_cast<bool>(tx->getTimeBounds());
+    uint64_t adjustedCloseTimeUpper = upperBoundCloseTimeOffset;
+    uint64_t adjustedCloseTimeLower = lowerBoundCloseTimeOffset;
+    if (!timeDependent)
+    {
+        adjustedCloseTimeUpper = 0;
+        adjustedCloseTimeLower = 0;
+    }
+
+    auto key = TxValidityKey{
+        mApp.getLedgerManager().getLastClosedLedgerHeader().hash,
+        tx->getFullHash(), adjustedCloseTimeLower, adjustedCloseTimeUpper};
+
+    if (mTxValidCache.exists(key))
+    {
+        mSCPMetrics.mTxValidityCacheHits.inc();
+        auto res = mTxValidCache.get(key);
+        // #ifdef SCP_DEBUGGING
+        auto diagnosticEvents =
+            DiagnosticEventManager::createForValidation(mApp.getConfig());
+        releaseAssert(tx->checkValid(mApp.getAppConnector(), ls, 0,
```

Comment:
I don't think this assertion will always hold. It's expected that sometimes the tx validity would fluctuate if the ledger changes between the validations. I believe it's fine, but in any case we shouldn't have this assertion and we should log something instead.

Reply:
I think the hash is safer to ensure we're actually validating against the correct state though, so I'd prefer to keep it that way (this is also consistent with the tx set cache)